### PR TITLE
Nonlinear least squares classes

### LINKED ISSUE #671: Expand cost function class for different NLLS models
Create a cost function class which implements `eval_cost` for least squares problems and then the inherit from that class to implement `eval_f` for the following models:

1. non-weighted NLLS `r(x) = y - f(x,p)`
2. weighted NLLS `r(x) = (y - f(x,p))/e`)`
2. square root NLLS `r(x) = y^(1/2) - f(x,p)^(1/2)`

### PR DESCRIPTION
**Pull request by wathen**
#### Description of Work

Fixes #671 

Adds new nonlinear least squares cost functions and removes the `use_errors` option from docs. Together with adding the new cost function classes I have updated the results index pages so that they explicitly say what cost function is being benchmarked.

#### Testing Instructions

1. Check that there are no instances of `options.use_errors`
2. Check that the three cost functions are implemented correctly
3. Check that the analytic Jacobian is correct for the `root_nlls` option

Function: Does the change do what it's supposed to?

Tests: Does it pass? Is there adequate coverage for new code?

Style: Is the coding style consistent? Is anything overly confusing?

Documentation: Is there a suitable change to documentation for this change?

### CODE REVIEW

File: docs/source/users/cost_func.rst [hunk 1]
```diff
@@ -0,0 +1,25 @@
+.. _cost_func:
+
+##############
+Cost functions
+##############
+
+Fitbenchmarking supports multiple different nonlinear least squares cost functions. These can be set via the `cost_func_type` option in :ref:`fitting_option`.
```
Comment (wathen) [inline docs/source/users/cost_func.rst:7]:
```suggestion
Fitbenchmarking supports multiple different nonlinear least squares cost functions. These can be set via the ``cost_func_type`` option in :ref:`fitting_option`.
```

File: fitbenchmarking/jacobian/analytic_jacobian.py [hunk 2]
```diff
@@ -21,7 +21,7 @@ def __init__(self, cost_func):
 
     def eval(self, params, **kwargs):
         """
-        Evaluates Jacobian of problem.eval_model or a weighted problem.eval_model
+        Evaluates Jacobian of cost_func.eval_cost
```
Comment (wathen) [inline fitbenchmarking/jacobian/analytic_jacobian.py:24]:
```suggestion
        Evaluates Jacobian of problem.eval_model
```

**Review by tyronerees — CHANGES_REQUESTED** [review]
This looks good.  

In addition to the one minor comment above, can we change `root_nlls_cost_func` to `hellinger_nlls_cost_func`?  This will link this cost function to the Hellinger distance cost function used in machine learning (see [here](https://stats.stackexchange.com/questions/154879/a-list-of-cost-functions-used-in-neural-networks-alongside-applications) or [here](http://www.tcs.tifr.res.in/~prahladh/teaching/2011-12/comm/lectures/l12.pdf)), which at least shows we haven't plucked it from thin air.

File: docs/source/users/options/fitting_option.rst [hunk 3]
```diff
@@ -120,12 +120,16 @@ Cost function (:code:`cost_func_type`)
 This sets the cost function to be used for the given data. Current cost
 functions supported are:
 
-* ``nlls`` - This sets the cost function to be weighted or non-weighted non-linear least squares, the weighting is controlled by the ``use_errors`` option.
+* ``nlls`` - This sets the cost function to be non-weighted non-linear least squares, :class:`~fitbenchmarking.cost_func.nlls_cost_func.NLLSCostFunc`.
+
+* ``weigthed_nlls`` - This sets the cost function to be weighted non-linear least squares, :class:`~fitbenchmarking.cost_func.weighted_nlls_cost_func.WeightedNLLSCostFunc`.
+
+* ``root_nlls`` - This sets the cost function to be root non-linear least squares, :class:`~fitbenchmarking.cost_func.root_nlls_cost_func.RootNLLSCostFunc`.
```
Comment (tyronerees) [inline docs/source/users/options/fitting_option.rst:127]:
```suggestion
* ``hellinger_nlls`` - This sets the cost function to be the Hellinger cost function, :class:`~fitbenchmarking.cost_func.root_nlls_cost_func.HellingerNLLSCostFunc`.
```

File: docs/source/users/options/fitting_option.rst [hunk 4]
```diff
@@ -120,12 +120,16 @@ Cost function (:code:`cost_func_type`)
...
 
 
 Default is ``nlls``
 
 .. code-block:: rst
 
     [FITTING]
-    cost_func_type: nlls
+    cost_func_type: root_nlls
```
Comment (tyronerees) [inline docs/source/users/options/fitting_option.rst:135]:
```suggestion
    cost_func_type: weighted_nlls
```

File: docs/source/users/options/fitting_option.rst [hunk 5]
```diff
@@ -120,12 +120,16 @@ Cost function (:code:`cost_func_type`)
 This sets the cost function to be used for the given data. Current cost
 functions supported are:
 
-* ``nlls`` - This sets the cost function to be weighted or non-weighted non-linear least squares, the weighting is controlled by the ``use_errors`` option.
+* ``nlls`` - This sets the cost function to be non-weighted non-linear least squares, :class:`~fitbenchmarking.cost_func.nlls_cost_func.NLLSCostFunc`.
+
+* ``weigthed_nlls`` - This sets the cost function to be weighted non-linear least squares, :class:`~fitbenchmarking.cost_func.weighted_nlls_cost_func.WeightedNLLSCostFunc`.
```
Comment (tyronerees) [inline docs/source/users/options/fitting_option.rst:125]:
```suggestion
* ``weighted_nlls`` - This sets the cost function to be weighted non-linear least squares, :class:`~fitbenchmarking.cost_func.weighted_nlls_cost_func.WeightedNLLSCostFunc`.
```
```suggestion
* ``weigthed_nlls`` - This sets the cost function to be weighted non-linear least squares, :class:`~fitbenchmarking.cost_func.weighted_nlls_cost_func.WeightedNLLSCostFunc`.
```

**Review by tyronerees — CHANGES_REQUESTED** [review]
Just a couple of doc changes needed, then all looks good to go.

**Review by tyronerees — APPROVED** [review]

**Pull request merged**